### PR TITLE
neverware: add an allow list for devices that can use SOF

### PR DESCRIPTION
**Pull request by nicholasbishop**
We want to prevent a big sudden QA burden, so limit SOF devices based
on system DMI info.

The allow list can be bypassed by adding this to kernel params:

    snd_intel_dspcfg.sof_bypass_allow_list=1

Note: originally I tried putting this DMI check at the top of
`sof_pci_probe`, but it didn't work correctly in the disallow case;
the output audio device would not show up in that case. Switched it to
`snd_intel_dsp_driver_probe` instead where it seems to work fine.

[OVER-13135]

autopick: master

[OVER-13135]: https://neverware.atlassian.net/browse/OVER-13135

### CODE REVIEW

**Comment by nicholasbishop** [comment]
@jefferymiller this is how I'm thinking of limiting SOF to particular devices that QA is ready to test on. (For now there's just one, I will add two more devices once I know what the DMI info for them looks like.)

I'm thinking of keeping all the SOF config options in https://github.com/neverware/chromiumos-overlay/pull/3036/files enabled, and just rely on this allow list to control what actually gets enabled. That way we don't have to sort through exactly which `CONFIG_` options are needed. And same for the firmware files; just include them all even if they aren't all used yet. What do you think of that approach?

**Comment by jefferymiller** [comment]
I like this approach. Enabling all the config options then limiting them with explicit workarounds.

**Comment by jefferymiller** [comment]
It might be nice here to have a module_param to avoid our check and allow a device. This way it might be easy to try a new model.
However, maybe that's only easy for a developer as we don't typically have users set module flags and a developer could just compile a new kernel.

**Comment by nicholasbishop** [comment]
Added the bypass param as suggested.

**Comment by nicholasbishop** [comment]
Added the ThinkPad X1 to the allow list, so I think all the expected changes are there now.

File: sound/hda/intel-dsp-config.c [hunk 1]
```diff
@@ -301,6 +301,57 @@ static int snd_intel_dsp_check_dmic(struct pci_dev *pci)
...
+};
+
+/* Neverware: add "snd_intel_dspcfg.sof_bypass_allow_list=1" to kernel
+ * params to disable the allow list. */
+static bool sof_bypass_allow_list = false;
+module_param(sof_bypass_allow_list, bool, S_IRUGO);
+MODULE_PARM_DESC(sof_bypass_allow_list,
+				 "Neverware: set to true to allow all compatible devices to use SOF");
+
+static int neverware_is_sof_allowed(void) {
```
Comment (jefferymiller) [inline sound/hda/intel-dsp-config.c:338]: nit: functions should have the `{` on the next line.

File: sound/hda/intel-dsp-config.c [hunk 2]
```diff
@@ -301,6 +301,57 @@ static int snd_intel_dsp_check_dmic(struct pci_dev *pci)
...
+	/* Allow the list to be bypassed */
+	if (sof_bypass_allow_list) {
+		printk(KERN_NOTICE "Neverware: SOF allow list disabled");
```
Comment (jefferymiller) [inline sound/hda/intel-dsp-config.c:341]: nit: It's probably better to use pr_notice macro here or even the dev_notice (would require the dev info though).

File: sound/hda/intel-dsp-config.c [hunk 3]
```diff
@@ -301,6 +301,57 @@ static int snd_intel_dsp_check_dmic(struct pci_dev *pci)
...
+		return 1;
+	}
+
+	if (dmi_check_system(neverware_sof_allowed_models)) {
+		printk(KERN_NOTICE "Neverware: allowing SOF");
+		return 1;
+	} else {
+		printk(KERN_WARNING "Neverware: disallowing SOF "
+			   "(may prevent audio devices from working)");
```
Comment (jefferymiller) [inline sound/hda/intel-dsp-config.c:350]: nit: Long lines aren't split in the kernel code.

**Review by jefferymiller — APPROVED** [review]
I only had some nitpicks.

File: sound/hda/intel-dsp-config.c (hunk 1 above)
Comment (nicholasbishop) [reply sound/hda/intel-dsp-config.c]: fixed

File: sound/hda/intel-dsp-config.c (hunk 3 above)
Comment (nicholasbishop) [reply sound/hda/intel-dsp-config.c]: fixed

File: sound/hda/intel-dsp-config.c (hunk 2 above)
Comment (nicholasbishop) [reply sound/hda/intel-dsp-config.c:342]: Any objection to leaving this as-is for simplicity?

Comment (jefferymiller) [reply sound/hda/intel-dsp-config.c:342]: no objection

**Pull request merged**